Add function is_quotient for matroids

danteluber opened the pull request:
Checks if two matroids of successive rank form a quotient.

danteluber committed:
matroid quotient stuff

lgoettgens commented on src/Combinatorics/Matroids/matroids.jl:
```diff
@@ -1019,3 +1019,48 @@ function automorphism_group(M::Matroid)
+
+@doc raw"""
+
+	is_quotient(Q1::Matroid,Q2::Matroid)
```
```suggestion
    is_quotient(Q1::Matroid, Q2::Matroid)
```
this needs 4 spaces of indentation

lgoettgens commented on src/Combinatorics/Matroids/matroids.jl:
```diff
@@ -1019,3 +1019,48 @@ function automorphism_group(M::Matroid)
+
+	is_quotient(Q1::Matroid,Q2::Matroid)
+
+Let Q1 and Q2 be matroids on the groundset $[n]$, with rank $rank(Q2) = rank(Q1)+1$.  Check if $Q1$ is a \emph{matroid quotient} of $Q2$.
```
```suggestion
Let Q1 and Q2 be matroids on the groundset $[n]$, with rank $rank(Q2) = rank(Q1)+1$.  Check if $Q1$ is a matroid quotient of $Q2$.
```
this does not support latex here

lgoettgens commented on src/Combinatorics/Matroids/matroids.jl:
```diff
@@ -1019,3 +1019,48 @@ function automorphism_group(M::Matroid)
+   
+    catch
+        
+        return false
```
Exceptions and try-catch blocks should not be used the intensely for regular functionality. In particular, you don't check the type of exception catched, this could be e.g. an `InterruptException` which is definitely not what you expect here.

danteluber committed:
changed quotient function

danteluber committed:
removed latex

danteluber committed:
typo in test

danteluber committed:
minor change in docs

danteluber committed:
export function

fingolfin commented on src/Combinatorics/Matroids/matroids.jl:
```diff
@@ -1019,3 +1019,31 @@ function automorphism_group(M::Matroid)
 end
+
+@doc raw"""
+    is_quotient(Q1::Matroid,Q2::Matroid)
```
Super nitpick, feel free to ignore
```suggestion
    is_quotient(Q1::Matroid, Q2::Matroid)
```

antonydellavecchia commented on src/Combinatorics/Matroids/matroids.jl:
```diff
@@ -1019,3 +1019,31 @@ function automorphism_group(M::Matroid)
+    B1 = [vcat(b,a) for b in bases(Q1)]
+    B2 = bases(Q2)
+    B = vcat(B1,B2)
+    M = Polymake.matroid.Matroid(BASES = B,N_ELEMENTS = a) 
```
is this not the same functions that you're using in the tests below? Better to just use the oscar function

antonydellavecchia commented on src/Combinatorics/Matroids/matroids.jl:
```diff
@@ -1019,3 +1019,31 @@ function automorphism_group(M::Matroid)
+    B2 = bases(Q2)
+    B = vcat(B1,B2)
+    M = Polymake.matroid.Matroid(BASES = B,N_ELEMENTS = a) 
+    return Polymake.matroid.check_basis_exchange_axiom(M.BASES)
```
similar for this, is there no Oscar equivalent for this? otherwise maybe it should exist?

fingolfin commented:
The matroids code is by @bschroter and @LukasKuehne so it would be good if they reviewed this PR. But of course if @antonydellavecchia or @benlorenz or anyone else has further thoughts, have a go as well.

But this has now been sitting for almost full month and it is unclear to me what the status is. Should we merge it? Should something be changed?

bschroter commented:
It is of course nice to have a function like this, but maybe call it `elementary quotient' or implement one of the general tests for arbitrary ranks.

More importantly there is an issue with the groundset. I don't like the unnecessary restriction to $[n]$, which is also never tested, as you are calling a polymake function. It would be better to ask for the same groundset.

danteluber committed:
simplifying function, no longer require specific groundset, can skip ranks

danteluber committed:
require same groundset

danteluber commented:
I have simplified the function and implemented @bschroter 's suggestions to make it more general. I just used the definition of a quotient  on page 2 [here](https://pdf.sciencedirectassets.com/272585/1-s2.0-S0001870823X00255/1-s2.0-S000187082300539X/main.pdf?X-Amz-Security-Token=IQoJb3JpZ2luX2VjEEAaCXVzLWVhc3QtMSJGMEQCIFq6uIFDkPHiLBkvSgofegalbE2skgP7FalfFX0HQoS%2FAiB8TfqEuGhmq%2B%2FOJ2Erru2QtaIjFAl0Kdw15USLnciKvCq7BQi5%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F8BEAUaDDA1OTAwMzU0Njg2NSIMnAi7Jhzcrdv4eXwlKo8Fa6upkWRkKbQ9%2Bsf77ljOaH0RttkSyj1PfaZPq9J6%2B2N7WLsPTAFrKKbe4dd95F25cghhwFqavE1IykYzKchsPKwpqyUiWzMC0Rh4wmsAm9jKtOuFN6GsSKzDpwAcn2xPASgpvl9uzfPVLij5MidOUwoDLMNU5YWgtYrph91rQcjWobyVM3QMUlBrQgJ86XfFjSwxxBXxPL8Lx1VO%2BZcfUVJ%2Bvda9wuXh7F%2Bd%2FurZZbjy9FGvKA7FVQ1u2n5A572qsQA9m2NIs%2Bv%2FW%2F1J%2BpkLqrdgvU8J3lAuwy%2BuQw%2BkI7QSlvC1UtZAOyK6zwIlMLusfDVpt8LdD8sLz%2FXUzC%2F%2BtGcNLHyhAsECrwTmt43LqKwutyaU1LMTjT%2F2m80igTPvzlIynBy2x3mvZnL%2BuuHmyoU3KzZIumUt53xsBps9Sjnz5VDo%2F8cDs6q%2BSvvEylbmzlONCTnkp14PN5uSv7pYwQaXpoYr01uGdO7H4NwmbZTggUXvxsLC3HqylP%2FEeZRzy19%2BdSmpxE1lgU2L3Ir1bfdk6gc7ZvvFKek%2BcopxlH1dSLwwIU4pS2HS1JdEeC6fwCv%2BTq%2F099qiGwVTuZ5CDM2rk%2FBMRWOefsIVhhDM80ljj050R3%2BzGv%2FhT7qEfWgm3GPQgI4C6wCL33zL6A2z7TdczoRkaZquxFkQAAYHM2BQ7l0ceMRd%2B4ZFzmZkvho58Mrttk00VkcSd1m0J4l1Ur%2Fvtkq8Rb5ahL0cfiASDZHFQslojlLIgQgj1fKOfVGmz%2BqPExuCHQwyWaDOmttrdbPuYI9G2j5p5OwJgT%2BtLpWvGItyG8FjtV0pV8TG8lEXsOXvhAP4O9lBAungU%2FkbOkdSjKvp%2BLmPF67CMhSbLzDt1suyBjqyAXlEdPR0PeCydMUMoIv0whdqZyVmyXpcv8ycRBLoigLona9Av3%2Fzomo17QxzKJUj4Hl4zhW9edXhy44wbufYd1U2QfHIueAQ0b25c16iV3gR9%2FaNARruS5HayyvMKdGPXZENLeRQsEo4O7YxuH9b%2F3FB%2FGTR0wsolI6FWdMeUCJMH2eOhfCG8aQb5fWWIWGJLPG4ciTfzMmwjdMqQNvzXTGGkALfsDfABNfaRZOJFcBOXY0%3D&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Date=20240526T082551Z&X-Amz-SignedHeaders=host&X-Amz-Expires=300&X-Amz-Credential=ASIAQ3PHCVTYWXFQOAUJ%2F20240526%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Signature=cbfbab4a4d141d90d56c79cad041dadf53caa2be5afab79a9975f4aa5411eb2e&hash=aa7c1122d646b46fc1f1d4844a7beaefb5e500bffdc2144e8c8b94c4a14403d9&host=68042c943591013ac2b2430a89b270f6af2c76d8dfd086a07176afe7c76c2c61&pii=S000187082300539X&tid=spdf-0ecddab8-768e-49df-8cb7-61b7d6f8cfc3&sid=fb565b5f978df343107b0b7305c693f133ccgxrqb&type=client&tsoh=d3d3LnNjaWVuY2VkaXJlY3QuY29t&ua=01025a565a0d57535907&rr=889c607ee8bc4480&cc=de)

danteluber committed:
changed @req message

bschroter commented on src/Combinatorics/Matroids/matroids.jl:
```diff
@@ -1019,3 +1019,28 @@ function automorphism_group(M::Matroid)
+@doc raw"""
+    is_quotient(Q1::Matroid,Q2::Matroid)
+
+Let Q1 and Q2 be matroids on the groundset $[n]$, with rank $rank(Q1) \leq rank(Q2)$.  
```
```suggestion
Let Q1 and Q2 be matroids on the same groundset, with rank $rank(Q1) \leq rank(Q2)$.  
```

bschroter commented on test/Combinatorics/Matroids/Matroids.jl:
I suggest to add a view more test cases.

1) One where the rank difference is at least 2, e.g. U_{4,4} and U_{2,4}
2) and 3) where the function fails because they trigger the two @req
4) One where the two matroids are not quotients

bschroter reviewed (commented):
Thank you, this already looks good. I suggest a few small changes.

danteluber committed:
more tests

danteluber committed:
fix test, change doc

fingolfin commented:
Is this good to go now? @bschroter your thoughts?

fingolfin commented on src/Combinatorics/Matroids/matroids.jl:
```diff
@@ -1019,3 +1019,28 @@ function automorphism_group(M::Matroid)
+"""
+function is_quotient(Q1::Matroid,Q2::Matroid)
+    @req matroid_groundset(Q1) == matroid_groundset(Q2) "matroids must be on same groundset" 
+    @req rank(Q1)<=rank(Q2) "matroids must be of equal or increasing rank"
```
Again, just wondering, but: why not this (and then also adjust the documentation comment) ?
```suggestion
    rank(Q1) <= rank(Q2) || return false
```
I mean, of course the rank condition is obviously necessary for Q1 to be a quotient of Q2, but we also don't raise an error if one checks `a < b` for two integers when `a` is positive and `b` negative, even though then "obviously" `a < b` is false.

(In contrast, having different groundsets kinda means "those two objects are not really comparable, so throwing an error there makes more sense to me).

Don't get me wrong, this is not a blocker for this PR, perhaps you are following established conventions in matroid theory -- it just seems odd from the outside, that's all :-)

danteluber replied:
I see what you are saying.

In my opinion, I could be wrong, telling the user that the matroids are not of increasing or equal rank can be informative and and reduces the potential for confusion about the output.

That is, if we have rank(Q1)<rank(Q2), and is_quotient(Q1,Q2) == true, the output is_quotient(Q2,Q1) == false can be misleading about the relationship between these matroids. 

Does this make sense? In any case, I am not attached to either approach.

fingolfin replied:
I have no strong reason to argue either way. Unless @bschroter disagrees, I'd say: let's leave it as is for now and if a user complaints about it later, we can still address it.

fingolfin commented on src/Combinatorics/Matroids/matroids.jl:
```diff
@@ -1019,3 +1019,28 @@ function automorphism_group(M::Matroid)
+    is_quotient(Q1::Matroid,Q2::Matroid)
+
+Let Q1 and Q2 be matroids on the same groundset with rank $rank(Q1) \leq rank(Q2)$.  
+Check if $Q1$ is a matroid quotient of $Q2$.
```
```suggestion
Let `Q1` and `Q2` be matroids on the same groundset with rank $rank(Q1) \leq rank(Q2)$.  
Check if `Q1` is a matroid quotient of `Q2`.
```

fingolfin commented on src/Combinatorics/Matroids/matroids.jl:
````diff
@@ -1019,3 +1019,28 @@ function automorphism_group(M::Matroid)
+true
+```
+"""
+function is_quotient(Q1::Matroid,Q2::Matroid)
````
Another super nitpick, but: why are these matroids called Q1 and Q2, and not, say M1 and M2 ? I could understand if they were named `Q` and `M` becayse the first one is tested for being a *Q*uotient and hence called *Q*.

But again: not a blocker, feel free to ignore, I am just wondering as a bystander ;-)

danteluber replied:
Fair point. I am in the habit of use Q for matroids, but not attached to it.

danteluber committed:
notation suggestions

fingolfin commented:
@bschroter @antonydellavecchia any further remarks? or can we merge this?

antonydellavecchia commented:
> @bschroter @antonydellavecchia any further remarks? or can we merge this?

Ok for me

bschroter commented:
Please go ahead. I think this can be merged.
